### PR TITLE
Signalable worker

### PR DESCRIPTION
**Pull request by dav-m85**
Worker ends nicely on USR1 signal.

### CODE REVIEW

File: Module/WorkerClass.php [hunk 1]
```diff
@@ -293,6 +301,14 @@ private function createJobCollection(ReflectionClass $reflectionClass, Reader $r
         return $jobCollection;
     }
 
+    private function loadSignalable(WorkAnnotation $workAnnotation, array $defaultSettings)
+    {
+
+        return  is_null($workAnnotation->signalable)
```
Comment (cotal) [inline Module/WorkerClass.php]: 2 spaces instead of one ?

File: Service/GearmanExecute.php [hunk 2]
```diff
@@ -111,9 +115,22 @@ private function callJob(Array $worker, \GearmanWorker $gearmanWorker = null)
             $this->addServers($gearmanWorker, $worker['servers']);
         }
 
+        // Signalable or not ?
+        if ($worker['signalable']) {
+            if( !extension_loaded('pcntl')){
+                throw new \Exception('pcntl should be loaded for signalable processes to work');
+            }
+            pcntl_signal(SIGUSR1,  array($this, "handleSignal"));
```
Comment (cotal) [inline Service/GearmanExecute.php]: simple quotes

File: Service/GearmanExecute.php [hunk 3]
```diff
@@ -193,27 +210,48 @@ private function runJob(\GearmanWorker $gearmanWorker, $objInstance, array $jobs
...
+                continue;
             }
 
-            /**
-             * Only finishes its execution if alive is false and iterations arrives to 0
-             */
-            if (!$alive && $iterations <= 0) {
+            if (! @$gearmanWorker->wait()) {
+                if ($gearmanWorker->returnCode() == GEARMAN_NO_ACTIVE_FDS)
+                {
```
Comment (cotal) [inline Service/GearmanExecute.php]: Checkstyle : not PSR2

File: Service/GearmanExecute.php [hunk 4]
```diff
@@ -193,27 +210,48 @@ private function runJob(\GearmanWorker $gearmanWorker, $objInstance, array $jobs
...
+                    sleep(5);
```
Comment (cotal) [inline Service/GearmanExecute.php:226]: why 5 ? doc ? comment ?

File: Service/GearmanExecute.php [hunk 5]
```diff
@@ -292,4 +336,21 @@ public function handleJob(\GearmanJob $job, $context)
...
     }
+
+    private $isRunning = false;
+
+    private $isRequestedToStop = false;
+
+    public function handleSignal($signo)
+    {
+        if($signo !== SIGUSR1) {
+
```
Comment (cotal) [inline Service/GearmanExecute.php]: useless blank line

File: Service/GearmanExecute.php [hunk 6]
```diff
@@ -193,27 +210,48 @@ private function runJob(\GearmanWorker $gearmanWorker, $objInstance, array $jobs
...
+                    break;
+                }
+
+                $iterations--;
+
+                /**
+                 * Only finishes its execution if alive is false and iterations arrives to 0, or if
+                 * signaled to stop
+                 */
+                if (!$alive && $iterations <= 0 && !$this->isRequestedToStop) {
```
Comment (cotal) [inline Service/GearmanExecute.php]: It seems to me that the code does not implement the documentation

What I read from the comment is :  
`if ((!$alive && $iterations <= 0) || $this->isRequestedToStop)`

Comment (dav-m85) [inline Service/GearmanExecute.php]: Correct. It is also duplicate code from l24 :)

File: Service/GearmanExecute.php [hunk 7]
```diff
@@ -78,76 +74,86 @@ public function executeJob($jobName, \GearmanWorker $gearmanWorker = null)
...
             $this->addServers($gearmanWorker, $worker['servers']);
         }
 
+        // Signalable or not ?
+        if ($worker['signalable']) {
+            if (!extension_loaded('pcntl')) {
+                throw new \Exception('pcntl should be loaded for signalable processes to work');
+            }
+            declare (ticks = 1);
+            pcntl_signal(SIGTERM,  array($this, "handleSignal"));
```
Comment (cotal) [inline Service/GearmanExecute.php]: simple quotes

**Comment by cotal** [comment]
OK

**Pull request merged**